File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        const int l = nums.size();
        if (l == 1) return nums[0];
        std::vector<int> val(l);
        val[0] = nums[0];
        val[1] = max(nums[0], nums[1]);
        for (int i = 2; i < l; ++i) {
            val[i] = max(val[i - 1], nums[i] + val[i - 2]);
        }
        return val[l - 1];
    }
};

File: 0198-house-robber/0198-house-robber.rs
impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        let n = nums.len();
        if n == 1 {
            return nums[0];
        }
        let mut dp = vec![0; n];
        dp[n - 1] = nums[n - 1];
        dp[n - 2] = nums[n - 1].max(nums[n - 2]);
        for i in (0..(n - 2)).rev() {
            dp[i] = (nums[i] + dp[i + 2]).max(dp[i + 1]);
        }
        dp[0]
    }
}